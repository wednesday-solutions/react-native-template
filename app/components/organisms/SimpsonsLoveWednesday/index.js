import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';
import { fonts } from '@themes';
import If from '@app/components/atoms/If';
import CharacterWithQuote from '@molecules/CharacterWithQuote';
import LogoWithInstructions from '@molecules/LogoWithInstructions';

const Err = styled.Text`
  ${fonts.style.standard()};
  text-align: center;
  margin-bottom: 5px;
  color: red;
`;

const SeparatedView = styled.View`
  > * {
    margin: 10px;
  }
`;
/**
 * A component that displays Simpsons-themed content related to Wednesday, including instructions, character details, and error messages.
 * @param {object} props - The props object containing component properties.
 * @param {string} props.instructions - The instructions text to display.
 * @param {object} props.user - The user object representing character details and quote.
 * @param {string} props.userErrorMessage - The error message to display if user data retrieval fails.
 * @returns {React.ReactNode} JSX elements displaying Simpsons-themed content based on provided props.
 */
function SimpsonsLoveWednesday({ instructions, user, userErrorMessage }) {
  return (
    <>
      <LogoWithInstructions instructions={instructions} />
      <If
        condition={!userErrorMessage}
        otherwise={<Err>{userErrorMessage}</Err>}
      >
        <SeparatedView>
          <CharacterWithQuote user={user} />
        </SeparatedView>
      </If>
    </>
  );
}

SimpsonsLoveWednesday.propTypes = {
  user: PropTypes.object,
  instructions: PropTypes.string,
  userErrorMessage: PropTypes.string
};

export default SimpsonsLoveWednesday;
