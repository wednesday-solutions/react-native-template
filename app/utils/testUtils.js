import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from 'app/rootReducer';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@atoms/LanguageProvider';

export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});
export const renderWithIntl = (children, renderFunction = render) =>
  renderFunction(
    <IntlProvider
      locale={DEFAULT_LOCALE}
      messages={translationMessages[DEFAULT_LOCALE]}
    >
      {children}
    </IntlProvider>
  );

export const renderProvider = children => {
  const { store } = createStore();
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        {children}
      </ConnectedLanguageProvider>
    </Provider>
  );
};

export const timeout = ms => new Promise(resolve => setTimeout(resolve, ms));
