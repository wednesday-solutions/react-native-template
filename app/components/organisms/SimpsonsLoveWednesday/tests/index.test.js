/**
 *
 * Tests for SimpsonsLoveWednesday
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import { rerender } from '@testing-library/react-native';
import SimpsonsLoveWednesday from '../index';

describe('<SimpsonsLoveWednesday />', () => {
  it('Should render and match the snapshot', () => {
    const baseElement = renderWithIntl(<SimpsonsLoveWednesday />);
    expect(baseElement).toMatchSnapshot();
  });
  it('Should render the Error component if userErrorMessage is not empty', () => {
    const props = {
      userErrorMessage: 'Fetch failed',
      instructions: 'PRESS CMD + D for iOS',
      user: {
        character: 'Homer',
        image:
          'https://www.onthisday.com/images/people/homer-simpson-medium.jpg',
        quote: "D'Oh!"
      }
    };
    const { getByText } = renderWithIntl(<SimpsonsLoveWednesday {...props} />);
    expect(getByText(props.userErrorMessage)).toBeTruthy();
    const updatedPropWithNullError = { ...props, userErrorMessage: null };
    const { getByText: textQueryOnReRender } = renderWithIntl(
      <SimpsonsLoveWednesday {...updatedPropWithNullError} />,
      rerender
    );
    expect(
      textQueryOnReRender(`${props.user.character} loves Wednesday`)
    ).toBeTruthy();
  });
});
