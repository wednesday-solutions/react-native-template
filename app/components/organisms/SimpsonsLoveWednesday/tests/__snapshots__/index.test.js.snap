// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpsonsLoveWednesday /> Should render and match the snapshot 1`] = `
Array [
  <View>
    <View
      style={
        Object {
          "height": 150,
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../app/assets/images/wednesday-logo.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 17,
            "fontStyle": "italic",
            "fontWeight": "normal",
            "marginBottom": 5,
            "textAlign": "center",
          },
        ]
      }
    />
  </View>,
  <View
    style={
      Array [
        Object {},
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 17,
            "fontWeight": "normal",
            "marginBottom": 5,
            "textAlign": "center",
          },
        ]
      }
    >
      character loves Wednesday
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 17,
            "fontWeight": "normal",
            "marginBottom": 5,
            "textAlign": "center",
          },
        ]
      }
    >
      because
    </Text>
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Array [
          Object {
            "height": 80,
            "marginBottom": 0,
            "marginLeft": "auto",
            "marginRight": "auto",
            "marginTop": 0,
            "width": 80,
          },
        ]
      }
      testID="character-image"
    />
    <Text
      style={
        Array [
          Object {
            "fontSize": 17,
            "fontWeight": "normal",
            "marginBottom": 5,
            "textAlign": "center",
          },
        ]
      }
    />
  </View>,
]
`;
